src/workerd/api/container.c++

```diff
@@ -48,18 +48,64 @@ void Container::start(jsg::Lock& js, jsg::Optional<StartupOptions> maybeOptions)
   }
 
   IoContext::current().addTask(req.sendIgnoringResult());
-
   running = true;
+  monitorOnBackgroundIfNeeded();
+}
+
+void Container::monitorOnBackgroundIfNeeded() {
+  if (monitoring || !running) {
+    return;
+  }
+  monitoring = true;
+
+  if (monitorKjPromise == kj::none) {
+    monitorKjPromise = rpcClient->monitorRequest(capnp::MessageSize{4, 0})
+                           .send()
+                           .then([](auto&& results) -> uint8_t {
+      return results.getExitCode();
+    }).fork();
+  }
+
+  auto req = KJ_ASSERT_NONNULL(monitorKjPromise)
+                 .addBranch()
+                 .then([](uint8_t exitCode) {}, [this, self = JSG_THIS](kj::Exception&& error) {
+    // Check if there is an existing monitor request made by the user.
+    // If not, we can abort the current context and error.
+    if (!monitoringExplicitly) {
+      IoContext::current().abort(kj::mv(error));
+    }
+  }).attach(kj::defer([this, self = JSG_THIS]() {
+    running = false;
```

Comment:
Here, too, you don't have the isolate lock, so accessing the JSG object is unsafe. Perhaps these flags can also be moved to the separate refcounted object.

Or, alternatively, perhaps updating these flags can be done in a continuation of the JSG promise returned by `monitor()`, instead of here. Note that if `monitor()` wasn't called, then we will have aborted the `IoContext` anyway, in which case updating these flags is not important.